flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/plan/rules/logical/PushWatermarkIntoTableSourceScanRuleTest.java

```diff
@@ -0,0 +1,188 @@
+				"  a int,\n" +
+				"  b bigint,\n" +
+				"  c timestamp(3),\n" +
+				"  d as c + interval '5' second,\n" +
+				"  watermark for d as d - interval '5' second\n" +
+				") WITH (\n" +
+				" 'connector' = 'values',\n" +
+				" 'table-source-class' = 'WATERMARK_PUSH_DOWN',\n" +
+				" 'bounded' = 'false'\n" +
+				")";
+		util.tableEnv().executeSql(ddl);
+		util.verifyPlan("select * from MyTable");
+	}
+
+	@Test
+	public void testSimpleTransposeNotNull() {
+		String ddl = "create table MyTable(" +
+				"  a int,\n" +
+				"  b bigint,\n" +
+				"  c timestamp(3) not null,\n" +
+				"  d as c + interval '5' second,\n" +
+				"  watermark for d as d - interval '5' second\n" +
+				") WITH (\n" +
+				" 'connector' = 'values',\n" +
+				" 'table-source-class' = 'WATERMARK_PUSH_DOWN',\n" +
+				" 'bounded' = 'false'\n" +
+				")";
+		util.tableEnv().executeSql(ddl);
+		util.verifyPlan("select * from MyTable");
+	}
+
+	@Test
+	public void testComputedColumnWithMultipleInputs() {
+		String ddl = "create table MyTable(" +
+				"  a string,\n" +
+				"  b string,\n" +
+				"  c as to_timestamp(a, b),\n" +
+				"  watermark for c as c - interval '5' second\n" +
+				") WITH (\n" +
+				" 'connector' = 'values',\n" +
+				" 'table-source-class' = 'WATERMARK_PUSH_DOWN',\n" +
+				" 'bounded' = 'false'\n" +
+				")";
+		util.tableEnv().executeSql(ddl);
+		util.verifyPlan("select * from MyTable");
+	}
+
+	@Test
+	public void testTransposeWithRow() {
+		String ddl = "create table MyTable(" +
+				"  a int,\n" +
+				"  b bigint,\n" +
+				"  c row<name string, d timestamp(3)>," +
+				"  e as c.d," +
+				"  watermark for e as e - interval '5' second\n" +
+				") WITH (\n" +
+				" 'connector' = 'values',\n" +
+				" 'table-source-class' = 'WATERMARK_PUSH_DOWN',\n" +
+				" 'bounded' = 'false'\n" +
+				")";
+		util.tableEnv().executeSql(ddl);
+		util.verifyPlan("select * from MyTable");
+	}
+
+	@Test
+	public void testTransposeWithNestedRow() {
+		String ddl = "create table MyTable(" +
+				"  a int,\n" +
+				"  b bigint,\n" +
+				"  c row<name string, d row<e string, f timestamp(3)>>," +
+				"  g as c.d.f," +
+				"  watermark for g as g - interval '5' second\n" +
+				") WITH (\n" +
+				" 'connector' = 'values',\n" +
+				" 'table-source-class' = 'WATERMARK_PUSH_DOWN',\n" +
+				" 'bounded' = 'false'\n" +
+				")";
+		util.tableEnv().executeSql(ddl);
+		util.verifyPlan("select * from MyTable");
+	}
+
+	@Test
+	public void testTransposeWithUdf() {
+		util.addFunction("func1", new InnerUdf());
+		String ddl = "create table MyTable(" +
+				"  a int,\n" +
+				"  b bigint,\n" +
+				"  c timestamp(3)," +
+				"  d as func1(c)," +
+				"  watermark for d as d - interval '5' second\n" +
+				") WITH (\n" +
+				" 'connector' = 'values',\n" +
+				" 'table-source-class' = 'WATERMARK_PUSH_DOWN',\n" +
+				" 'bounded' = 'false'\n" +
+				")";
+		util.tableEnv().executeSql(ddl);
+		util.verifyPlan("select * from MyTable");
```

Comment:
add some tests about 
1. projection/filter in `select`, 
2. watermark expression contains two field references

Reply:
Please refer to `testWatermarkOnComputedColumnWithQuery` 
 and `testWatermarkOnComputedColumnWithMultipleInputs`.